[ADD] Documentation update in `base_trainer.py`

theodorju opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?
<!-- 
* [ ] Have you followed the guidelines in our Contributing document?
-->


## Description
<!--- Describe your changes in detail -->
I have updated the documentation of `base_trainer` for #406 in places where it seemed reasonable to add documentation.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Changes fix #406 
## How has this been tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
Since the changes are only in the documentation, I have not executed any tests for this issue.

ravinkohli commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -44,14 +44,28 @@ def __init__(self,
 
+        Returns:
+            True if the current epoch is larger than the maximum epochs, False otherwise.
+            Additionally, returns False if the run is without this constrain.
```
```suggestion
            Additionally, returns False if the run is without this constraint.
```

ravinkohli commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -60,6 +74,13 @@ def is_max_epoch_reached(self, epoch: int) -> bool:
+            runtime is reached.
+        Returns:
+            True if the maximum runtime is reached, False otherwise.
+            Additionally, returns False if the run is without this constrain.
```
```suggestion
            Additionally, returns False if the run is without this constraint.
```

ravinkohli commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -44,14 +44,28 @@ def __init__(self,
         It also allows to define a 'epoch_or_time' budget type, which means,
         the first of them both which is exhausted, is honored
+        Args:
+            budget_type (str): Type of budget to be used when fitting the pipeline.
```
Could you add a new line and indentation to the description of `budget_type`? You can check the base_task.py file for how we format the docstrings.

ravinkohli commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -44,14 +44,28 @@ def __init__(self,
+        Args:
+            budget_type (str): Type of budget to be used when fitting the pipeline.
+                Possible values are 'epochs', 'runtime', or 'epoch_or_time'
+            max_epochs (Optional[int], default=None): Maximum number of epochs to train the pipeline for
```
Could you also do this for all the other parameter descriptions?

ravinkohli commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -118,8 +144,20 @@ def add_performance(self,
+        Args:
+            epoch (int): the current epoch
+            start_time (float): timestamp at the beginning of current epoch
+            end_time (float): timestamp when gathering the information
```
```suggestion
            end_time (float): timestamp when gathering the information after the current epoch.
```

ravinkohli reviewed (changes requested):
Hey, thanks for the pull request. I have left some minor comments after which I think we can merge this PR.

theodorju commented:
Thanks for the comments. I have applied all the suggested changes. Additionally:
- I have kept the 120 max line width, and also modified some pre-existing docstrings accordingly
- Added more detailed descriptions in return sections, where applicable
- Updated some docstrings that I've missed in the previous commit (e.g. [`criterion_preparation`](https://github.com/theodorju/Auto-PyTorch/blob/953b3a086b81d40a7fc26a9621f521df4732eb78/autoPyTorch/pipeline/components/training/trainer/base_trainer.py#L708))
